src/persist/EpochSys.hpp

```diff
@@ -13,281 +13,35 @@
-        }
-        void on_end_transaction(EpochSys* esys, uint64_t c){
-            uint64_t curr_cnt = trans_cnt.fetch_add(1, std::memory_order_acq_rel);
-            if (curr_cnt % epoch_threshold == 0){
-                esys->advance_epoch(c);
-            } else if (curr_cnt % help_threshold == 0){
-                esys->help_local();
-            }
-        }
-    };
-
-    class DedicatedEpochAdvancer : public EpochAdvancer{
-        EpochSys* esys;
-        std::thread advancer_thread;
-        std::atomic<bool> started;
-        uint64_t epoch_length = 100*1000;
-        void advancer(){
-            while(!started.load()){}
-            while(started.load()){
-                esys->advance_epoch_dedicated();
-                std::this_thread::sleep_for(std::chrono::microseconds(epoch_length));
-            }
-            // std::cout<<"advancer_thread terminating..."<<std::endl;
-        }
-    public:
-        DedicatedEpochAdvancer(GlobalTestConfig* gtc, EpochSys* es):esys(es){
-            if (gtc->checkEnv("EpochLength")){
-                epoch_length = stoi(gtc->getEnv("EpochLength"));
-            } else {
-                epoch_length = 100*1000;
-            }
-            if (gtc->checkEnv("EpochLengthUnit")){
-                std::string env_unit = gtc->getEnv("EpochLengthUnit");
-                if (env_unit == "Second"){
-                    epoch_length *= 1000000;
-                } else if (env_unit == "Millisecond"){
-                    epoch_length *= 1000;
-                } else if (env_unit == "Microsecond"){
-                    // do nothing.
-                } else {
-                    errexit("time unit not supported.");
-                }
-            }
-
-            started.store(false);
-            advancer_thread = std::move(std::thread(&DedicatedEpochAdvancer::advancer, this));
-            started.store(true);
-        }
-        ~DedicatedEpochAdvancer(){
-            // std::cout<<"terminating advancer_thread"<<std::endl;
-            started.store(false);
-            if (advancer_thread.joinable()){
-                advancer_thread.join();
-            }
-            // std::cout<<"terminated advancer_thread"<<std::endl;
-        }
-        void set_epoch_freq(int epoch_interval){
-        }
-        void set_help_freq(int help_interval){
-        }
-        void on_end_transaction(EpochSys* esys, uint64_t c){
-            // do nothing here.
-        }
-    };
-
-    class NoEpochAdvancer : public EpochAdvancer{
-        // an epoch advancer that does absolutely nothing.
-    public:
-        // GlobalCounterEpochAdvancer();
-        void set_epoch_freq(int epoch_power){}
-        void set_help_freq(int help_power){}
-        void on_end_transaction(EpochSys* esys, uint64_t c){}
-    };
-
-    friend class EpochSysVerifyTest;
-    
 private:
     // persistent fields:
     Epoch* epoch_container = nullptr;
     std::atomic<uint64_t>* global_epoch = nullptr;
 
+    // semi-persistent fields:
```

Comment:
This is an interesting problem. UID generator needs to be persisted _somehow_ (we don't want it to generate existing IDs after recovery), but it doesn't worth to be persisted _precisely_, since persisting on every UID generation eats up the performance advantage of buffered durability.
The idea here is we persist a "boundary" every 2^x IDs, for all IDs smaller than that boundary. After a crash-recovery, we continue from that boundary.

Reply:
Cool. Thanks. Setting this conversation resolved.